File: problems/basic/20/solution.py
# Accepted: 34ms (83.79%), 17.45MB (6.15%)

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for c in s:
            if c in "({[":
                stack.append(c)
                continue

            # return false if unmatched closing parenthesis
            if len(stack) == 0:
                return False

            pop = stack.pop()

            if c == ")" and pop != "(":
                return False
            if c == "}" and pop != "{":
                return False
            if c == "]" and pop != "[":
                return False

        # return false if any unmatched opening parenthesis
        if len(stack) != 0:
            return False

        return True

File: problems/basic/20/solution.ts
// Accepted: 60ms (60.80%), 44.60 (49.60%)

function isValid(s: string): boolean {
    const stack: string[] = [];

    for (const c of s) {
        if ("({[".includes(c)) {
            stack.push(c);
            continue;
        }

        // return false if unmatched closing parenthesis
        if (stack.length === 0) {
            return false;
        }

        const pop = stack.pop();

        if (c === ")" && pop !== "(") {
            return false;
        }
        if (c === "}" && pop !== "{") {
            return false;
        }
        if (c === "]" && pop !== "[") {
            return false;
        }
    }

    // return false if any unmatched opening parenthesis
    if (stack.length !== 0) {
        return false;
    }

    return true;
};
